/*
 * BrAPI-Germplasm
 * The Breeding API (BrAPI) is a Standardized REST ful Web Service API Specification for communicating Plant Breeding Data. BrAPI allows for easy data sharing between databases and tools involved in plant breeding. <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">General Reference Documentation</h2> <div class=\"gen-info-link\"><a href=\"https://wiki.brapi.org/index.php/RESTfulness\">URL Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://wiki.brapi.org/index.php/Response_Structure\">Response Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://wiki.brapi.org/index.php/Dates_and_Times\">Date/Time Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://wiki.brapi.org/index.php/Location_Coordinates\">Location Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://wiki.brapi.org/index.php/Error_Handling\">Error Handling</a></div> <div class=\"gen-info-link\"><a href=\"https://wiki.brapi.org/index.php/Search_Services\">Search Services</a></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Core</h2> <div class=\"brapi-section-description\">The BrAPI Core module contains high level entities used for organization and management. This includes Programs, Trials, Studies, Locations, People, and Lists</div> <div class=\"version-number\">V2.1</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/brapi-V2.1/Specification/BrAPI-Core\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Core/2.1\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapicore21.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Phenotyping</h2> <div class=\"brapi-section-description\">The BrAPI Phenotyping module contains entities related to phenotypic observations. This includes Observation Units, Observations, Observation Variables, Traits, Scales, Methods, and Images</div> <div class=\"version-number\">V2.1</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/brapi-V2.1/Specification/BrAPI-Phenotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Phenotyping/2.1\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapiphenotyping21.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Genotyping</h2> <div class=\"brapi-section-description\">The BrAPI Genotyping module contains entities related to genotyping analysis. This includes Samples, Markers, Variant Sets, Variants, Call Sets, Calls, References, Reads, and Vendor Orders</div> <div class=\"version-number\">V2.1</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/brapi-V2.1/Specification/BrAPI-Genotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Genotyping/2.1\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigenotyping21.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"current-brapi-section brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Germplasm</h2> <div class=\"brapi-section-description\">The BrAPI Germplasm module contains entities related to germplasm management. This includes Germplasm, Germplasm Attributes, Seed Lots, Crosses, Pedigree, and Progeny</div> <div class=\"version-number\">V2.1</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/brapi-V2.1/Specification/BrAPI-Germplasm\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Germplasm/2.1\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigermplasm21.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <style> .link-btn{ float: left;  margin: 2px 10px 0 0;  padding: 0 5px;  border-radius: 5px;  background-color: #ddd; } .stop-float{   clear: both; } .version-number{   float: left;    margin: 5px 10px 0 5px; } .brapi-section-title{   margin: 0 10px 0 0;   font-size: 20px; } .current-brapi-section{   font-weight: bolder;   border-radius: 5px;    background-color: #ddd; } .brapi-section{   padding: 5px 5px;  } .brapi-section-description{   margin: 5px 0 0 5px; } </style>
 *
 * OpenAPI spec version: 2.1
 *
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 */

package org.brapi.model.v21.germplasm;

import com.google.gson.TypeAdapter;
import com.google.gson.annotations.JsonAdapter;
import com.google.gson.annotations.SerializedName;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;
import io.swagger.v3.oas.annotations.media.Schema;
import org.threeten.bp.LocalDate;

import java.io.IOException;
import java.util.*;

/**
 * GermplasmNewRequest
 */

@javax.annotation.Generated(value = "io.swagger.codegen.v3.generators.java.JavaClientCodegen", date = "2022-10-25T17:45:22.672Z[GMT]")
public class GermplasmNewRequest {
    @SerializedName("accessionNumber")
    private String accessionNumber = null;

    @SerializedName("acquisitionDate")
    private LocalDate acquisitionDate = null;

    @SerializedName("additionalInfo")
    private Map<String, String> additionalInfo = null;

    /**
     * MCPD (v2.1) (SAMPSTAT) 19. The coding scheme proposed can be used at 3 different levels of detail: either by using the general codes such as 100, 200, 300, 400, or by using the more specific codes such as 110, 120, etc.   100) Wild  110) Natural  120) Semi-natural/wild  130) Semi-natural/sown  200) Weedy  300) Traditional cultivar/landrace  400) Breeding/research material  410) Breeders line  411) Synthetic population  412) Hybrid  413) Founder stock/base population  414) Inbred line (parent of hybrid cultivar)  415) Segregating population  416) Clonal selection  420) Genetic stock  421) Mutant (e.g. induced/insertion mutants, tilling populations)  422) Cytogenetic stocks (e.g. chromosome addition/substitution, aneuploids,  amphiploids)  423) Other genetic stocks (e.g. mapping populations)  500) Advanced or improved cultivar (conventional breeding methods)  600) GMO (by genetic engineering)  999) Other (Elaborate in REMARKS field)
     */
    @JsonAdapter(BiologicalStatusOfAccessionCodeEnum.Adapter.class)
    public enum BiologicalStatusOfAccessionCodeEnum {
        _100("100"),
        _110("110"),
        _120("120"),
        _130("130"),
        _200("200"),
        _300("300"),
        _400("400"),
        _410("410"),
        _411("411"),
        _412("412"),
        _413("413"),
        _414("414"),
        _415("415"),
        _416("416"),
        _420("420"),
        _421("421"),
        _422("422"),
        _423("423"),
        _500("500"),
        _600("600"),
        _999("999");

        private String value;

        BiologicalStatusOfAccessionCodeEnum(String value) {
            this.value = value;
        }

        public String getValue() {
            return value;
        }

        @Override
        public String toString() {
            return String.valueOf(value);
        }

        public static BiologicalStatusOfAccessionCodeEnum fromValue(String input) {
            for (BiologicalStatusOfAccessionCodeEnum b : BiologicalStatusOfAccessionCodeEnum.values()) {
                if (b.value.equals(input)) {
                    return b;
                }
            }
            return null;
        }

        public static class Adapter extends TypeAdapter<BiologicalStatusOfAccessionCodeEnum> {
            @Override
            public void write(final JsonWriter jsonWriter, final BiologicalStatusOfAccessionCodeEnum enumeration) throws IOException {
                jsonWriter.value(String.valueOf(enumeration.getValue()));
            }

            @Override
            public BiologicalStatusOfAccessionCodeEnum read(final JsonReader jsonReader) throws IOException {
                Object value = jsonReader.nextString();
                return BiologicalStatusOfAccessionCodeEnum.fromValue((String) (value));
            }
        }
    }

    @SerializedName("biologicalStatusOfAccessionCode")
    private BiologicalStatusOfAccessionCodeEnum biologicalStatusOfAccessionCode = null;

    @SerializedName("biologicalStatusOfAccessionDescription")
    private String biologicalStatusOfAccessionDescription = null;

    @SerializedName("breedingMethodDbId")
    private String breedingMethodDbId = null;

    @SerializedName("breedingMethodName")
    private String breedingMethodName = null;

    @SerializedName("collection")
    private String collection = null;

    @SerializedName("commonCropName")
    private String commonCropName = null;

    @SerializedName("countryOfOriginCode")
    private String countryOfOriginCode = null;

    @SerializedName("defaultDisplayName")
    private String defaultDisplayName = null;

    @SerializedName("documentationURL")
    private String documentationURL = null;

    @SerializedName("donors")
    private List<GermplasmDonors> donors = null;

    @SerializedName("externalReferences")
    private List<CrossExternalReferences> externalReferences = null;

    @SerializedName("genus")
    private String genus = null;

    @SerializedName("germplasmName")
    private String germplasmName = null;

    @SerializedName("germplasmOrigin")
    private List<GermplasmGermplasmOrigin> germplasmOrigin = null;

    @SerializedName("germplasmPUI")
    private String germplasmPUI = null;

    @SerializedName("germplasmPreprocessing")
    private String germplasmPreprocessing = null;

    @SerializedName("instituteCode")
    private String instituteCode = null;

    @SerializedName("instituteName")
    private String instituteName = null;

    @SerializedName("pedigree")
    private String pedigree = null;

    @SerializedName("seedSource")
    private String seedSource = null;

    @SerializedName("seedSourceDescription")
    private String seedSourceDescription = null;

    @SerializedName("species")
    private String species = null;

    @SerializedName("speciesAuthority")
    private String speciesAuthority = null;

    @SerializedName("storageTypes")
    private List<GermplasmStorageTypes> storageTypes = null;

    @SerializedName("subtaxa")
    private String subtaxa = null;

    @SerializedName("subtaxaAuthority")
    private String subtaxaAuthority = null;

    @SerializedName("synonyms")
    private List<GermplasmSynonyms> synonyms = null;

    @SerializedName("taxonIds")
    private List<GermplasmTaxonIds> taxonIds = null;

    public GermplasmNewRequest accessionNumber(String accessionNumber) {
        this.accessionNumber = accessionNumber;
        return this;
    }

    /**
     * The unique identifier for a material or germplasm within a genebank  MCPD (v2.1) (ACCENUMB) 2. This is the unique identifier for accessions within a genebank, and is assigned when a sample is entered into the genebank collection (e.g. \&quot;PI 113869\&quot;).
     *
     * @return accessionNumber
     **/
    @Schema(example = "A0000003", description = "The unique identifier for a material or germplasm within a genebank  MCPD (v2.1) (ACCENUMB) 2. This is the unique identifier for accessions within a genebank, and is assigned when a sample is entered into the genebank collection (e.g. \"PI 113869\").")
    public String getAccessionNumber() {
        return accessionNumber;
    }

    public void setAccessionNumber(String accessionNumber) {
        this.accessionNumber = accessionNumber;
    }

    public GermplasmNewRequest acquisitionDate(LocalDate acquisitionDate) {
        this.acquisitionDate = acquisitionDate;
        return this;
    }

    /**
     * The date a material or germplasm was acquired by the genebank   MCPD (v2.1) (ACQDATE) 12. Date on which the accession entered the collection [YYYYMMDD] where YYYY is the year, MM is the month and DD is the day. Missing data (MM or DD) should be indicated with hyphens or \&quot;00\&quot; [double zero].
     *
     * @return acquisitionDate
     **/
    @Schema(description = "The date a material or germplasm was acquired by the genebank   MCPD (v2.1) (ACQDATE) 12. Date on which the accession entered the collection [YYYYMMDD] where YYYY is the year, MM is the month and DD is the day. Missing data (MM or DD) should be indicated with hyphens or \"00\" [double zero].")
    public LocalDate getAcquisitionDate() {
        return acquisitionDate;
    }

    public void setAcquisitionDate(LocalDate acquisitionDate) {
        this.acquisitionDate = acquisitionDate;
    }

    public GermplasmNewRequest additionalInfo(Map<String, String> additionalInfo) {
        this.additionalInfo = additionalInfo;
        return this;
    }

    public GermplasmNewRequest putAdditionalInfoItem(String key, String additionalInfoItem) {
        if (this.additionalInfo == null) {
            this.additionalInfo = new HashMap<String, String>();
        }
        this.additionalInfo.put(key, additionalInfoItem);
        return this;
    }

    /**
     * A free space containing any additional information related to a particular object. A data source may provide any JSON object, unrestriced by the BrAPI specification.
     *
     * @return additionalInfo
     **/
    @Schema(description = "A free space containing any additional information related to a particular object. A data source may provide any JSON object, unrestriced by the BrAPI specification.")
    public Map<String, String> getAdditionalInfo() {
        return additionalInfo;
    }

    public void setAdditionalInfo(Map<String, String> additionalInfo) {
        this.additionalInfo = additionalInfo;
    }

    public GermplasmNewRequest biologicalStatusOfAccessionCode(BiologicalStatusOfAccessionCodeEnum biologicalStatusOfAccessionCode) {
        this.biologicalStatusOfAccessionCode = biologicalStatusOfAccessionCode;
        return this;
    }

    /**
     * MCPD (v2.1) (SAMPSTAT) 19. The coding scheme proposed can be used at 3 different levels of detail: either by using the general codes such as 100, 200, 300, 400, or by using the more specific codes such as 110, 120, etc.   100) Wild  110) Natural  120) Semi-natural/wild  130) Semi-natural/sown  200) Weedy  300) Traditional cultivar/landrace  400) Breeding/research material  410) Breeders line  411) Synthetic population  412) Hybrid  413) Founder stock/base population  414) Inbred line (parent of hybrid cultivar)  415) Segregating population  416) Clonal selection  420) Genetic stock  421) Mutant (e.g. induced/insertion mutants, tilling populations)  422) Cytogenetic stocks (e.g. chromosome addition/substitution, aneuploids,  amphiploids)  423) Other genetic stocks (e.g. mapping populations)  500) Advanced or improved cultivar (conventional breeding methods)  600) GMO (by genetic engineering)  999) Other (Elaborate in REMARKS field)
     *
     * @return biologicalStatusOfAccessionCode
     **/
    @Schema(example = "420", description = "MCPD (v2.1) (SAMPSTAT) 19. The coding scheme proposed can be used at 3 different levels of detail: either by using the general codes such as 100, 200, 300, 400, or by using the more specific codes such as 110, 120, etc.   100) Wild  110) Natural  120) Semi-natural/wild  130) Semi-natural/sown  200) Weedy  300) Traditional cultivar/landrace  400) Breeding/research material  410) Breeders line  411) Synthetic population  412) Hybrid  413) Founder stock/base population  414) Inbred line (parent of hybrid cultivar)  415) Segregating population  416) Clonal selection  420) Genetic stock  421) Mutant (e.g. induced/insertion mutants, tilling populations)  422) Cytogenetic stocks (e.g. chromosome addition/substitution, aneuploids,  amphiploids)  423) Other genetic stocks (e.g. mapping populations)  500) Advanced or improved cultivar (conventional breeding methods)  600) GMO (by genetic engineering)  999) Other (Elaborate in REMARKS field)")
    public BiologicalStatusOfAccessionCodeEnum getBiologicalStatusOfAccessionCode() {
        return biologicalStatusOfAccessionCode;
    }

    public void setBiologicalStatusOfAccessionCode(BiologicalStatusOfAccessionCodeEnum biologicalStatusOfAccessionCode) {
        this.biologicalStatusOfAccessionCode = biologicalStatusOfAccessionCode;
    }

    public GermplasmNewRequest biologicalStatusOfAccessionDescription(String biologicalStatusOfAccessionDescription) {
        this.biologicalStatusOfAccessionDescription = biologicalStatusOfAccessionDescription;
        return this;
    }

    /**
     * Supplemental text description for &#x27;biologicalStatusOfAccessionCode&#x27;
     *
     * @return biologicalStatusOfAccessionDescription
     **/
    @Schema(example = "Genetic stock", description = "Supplemental text description for 'biologicalStatusOfAccessionCode'")
    public String getBiologicalStatusOfAccessionDescription() {
        return biologicalStatusOfAccessionDescription;
    }

    public void setBiologicalStatusOfAccessionDescription(String biologicalStatusOfAccessionDescription) {
        this.biologicalStatusOfAccessionDescription = biologicalStatusOfAccessionDescription;
    }

    public GermplasmNewRequest breedingMethodDbId(String breedingMethodDbId) {
        this.breedingMethodDbId = breedingMethodDbId;
        return this;
    }

    /**
     * The unique identifier for the breeding method used to create this germplasm
     *
     * @return breedingMethodDbId
     **/
    @Schema(example = "ffcce7ef", description = "The unique identifier for the breeding method used to create this germplasm")
    public String getBreedingMethodDbId() {
        return breedingMethodDbId;
    }

    public void setBreedingMethodDbId(String breedingMethodDbId) {
        this.breedingMethodDbId = breedingMethodDbId;
    }

    public GermplasmNewRequest breedingMethodName(String breedingMethodName) {
        this.breedingMethodName = breedingMethodName;
        return this;
    }

    /**
     * human readable name of the breeding method
     *
     * @return breedingMethodName
     **/
    @Schema(example = "Male Backcross", description = "human readable name of the breeding method")
    public String getBreedingMethodName() {
        return breedingMethodName;
    }

    public void setBreedingMethodName(String breedingMethodName) {
        this.breedingMethodName = breedingMethodName;
    }

    public GermplasmNewRequest collection(String collection) {
        this.collection = collection;
        return this;
    }

    /**
     * A specific panel/collection/population name this germplasm belongs to.
     *
     * @return collection
     **/
    @Schema(example = "Rice Diversity Panel 1 (RDP1)", description = "A specific panel/collection/population name this germplasm belongs to.")
    public String getCollection() {
        return collection;
    }

    public void setCollection(String collection) {
        this.collection = collection;
    }

    public GermplasmNewRequest commonCropName(String commonCropName) {
        this.commonCropName = commonCropName;
        return this;
    }

    /**
     * Common name for the crop   MCPD (v2.1) (CROPNAME) 10. Common name of the crop. Example: \&quot;malting barley\&quot;, \&quot;mas\&quot;.
     *
     * @return commonCropName
     **/
    @Schema(example = "Maize", required = true, description = "Common name for the crop   MCPD (v2.1) (CROPNAME) 10. Common name of the crop. Example: \"malting barley\", \"mas\".")
    public String getCommonCropName() {
        return commonCropName;
    }

    public void setCommonCropName(String commonCropName) {
        this.commonCropName = commonCropName;
    }

    public GermplasmNewRequest countryOfOriginCode(String countryOfOriginCode) {
        this.countryOfOriginCode = countryOfOriginCode;
        return this;
    }

    /**
     * 3-letter ISO 3166-1 code of the country in which the sample was originally collected   MCPD (v2.1) (ORIGCTY) 13. 3-letter ISO 3166-1 code of the country in which the sample was originally collected (e.g. landrace, crop wild relative, farmers variety), bred or selected (breeding lines, GMOs, segregating populations, hybrids, modern cultivars, etc.). Note- Descriptors 14 to 16 below should be completed accordingly only if it was \&quot;collected\&quot;.
     *
     * @return countryOfOriginCode
     **/
    @Schema(example = "BES", description = "3-letter ISO 3166-1 code of the country in which the sample was originally collected   MCPD (v2.1) (ORIGCTY) 13. 3-letter ISO 3166-1 code of the country in which the sample was originally collected (e.g. landrace, crop wild relative, farmers variety), bred or selected (breeding lines, GMOs, segregating populations, hybrids, modern cultivars, etc.). Note- Descriptors 14 to 16 below should be completed accordingly only if it was \"collected\".")
    public String getCountryOfOriginCode() {
        return countryOfOriginCode;
    }

    public void setCountryOfOriginCode(String countryOfOriginCode) {
        this.countryOfOriginCode = countryOfOriginCode;
    }

    public GermplasmNewRequest defaultDisplayName(String defaultDisplayName) {
        this.defaultDisplayName = defaultDisplayName;
        return this;
    }

    /**
     * Human readable name used for display purposes
     *
     * @return defaultDisplayName
     **/
    @Schema(example = "A0000003", description = "Human readable name used for display purposes")
    public String getDefaultDisplayName() {
        return defaultDisplayName;
    }

    public void setDefaultDisplayName(String defaultDisplayName) {
        this.defaultDisplayName = defaultDisplayName;
    }

    public GermplasmNewRequest documentationURL(String documentationURL) {
        this.documentationURL = documentationURL;
        return this;
    }

    /**
     * A URL to the human readable documentation of an object
     *
     * @return documentationURL
     **/
    @Schema(example = "https://wiki.brapi.org", description = "A URL to the human readable documentation of an object")
    public String getDocumentationURL() {
        return documentationURL;
    }

    public void setDocumentationURL(String documentationURL) {
        this.documentationURL = documentationURL;
    }

    public GermplasmNewRequest donors(List<GermplasmDonors> donors) {
        this.donors = donors;
        return this;
    }

    public GermplasmNewRequest addDonorsItem(GermplasmDonors donorsItem) {
        if (this.donors == null) {
            this.donors = new ArrayList<GermplasmDonors>();
        }
        this.donors.add(donorsItem);
        return this;
    }

    /**
     * List of donor institutes
     *
     * @return donors
     **/
    @Schema(description = "List of donor institutes")
    public List<GermplasmDonors> getDonors() {
        return donors;
    }

    public void setDonors(List<GermplasmDonors> donors) {
        this.donors = donors;
    }

    public GermplasmNewRequest externalReferences(List<CrossExternalReferences> externalReferences) {
        this.externalReferences = externalReferences;
        return this;
    }

    public GermplasmNewRequest addExternalReferencesItem(CrossExternalReferences externalReferencesItem) {
        if (this.externalReferences == null) {
            this.externalReferences = new ArrayList<CrossExternalReferences>();
        }
        this.externalReferences.add(externalReferencesItem);
        return this;
    }

    /**
     * An array of external reference ids. These are references to this piece of data in an external system. Could be a simple string or a URI.
     *
     * @return externalReferences
     **/
    @Schema(example = "[{\"referenceId\":\"doi:10.155454/12341234\",\"referenceSource\":\"DOI\"},{\"referenceId\":\"75a50e76\",\"referenceSource\":\"Remote Data Collection Upload Tool\"}]", description = "An array of external reference ids. These are references to this piece of data in an external system. Could be a simple string or a URI.")
    public List<CrossExternalReferences> getExternalReferences() {
        return externalReferences;
    }

    public void setExternalReferences(List<CrossExternalReferences> externalReferences) {
        this.externalReferences = externalReferences;
    }

    public GermplasmNewRequest genus(String genus) {
        this.genus = genus;
        return this;
    }

    /**
     * Genus name for taxon. Initial uppercase letter required.  MCPD (v2.1) (GENUS) 5. Genus name for taxon. Initial uppercase letter required.  MIAPPE V1.1 (DM-43) Genus - Genus name for the organism under study, according to standard scientific nomenclature.
     *
     * @return genus
     **/
    @Schema(example = "Aspergillus", description = "Genus name for taxon. Initial uppercase letter required.  MCPD (v2.1) (GENUS) 5. Genus name for taxon. Initial uppercase letter required.  MIAPPE V1.1 (DM-43) Genus - Genus name for the organism under study, according to standard scientific nomenclature.")
    public String getGenus() {
        return genus;
    }

    public void setGenus(String genus) {
        this.genus = genus;
    }

    public GermplasmNewRequest germplasmName(String germplasmName) {
        this.germplasmName = germplasmName;
        return this;
    }

    /**
     * Name of the germplasm. It can be the preferred name and does not have to be unique.  MCPD (v2.1) (ACCENAME) 11. Either a registered or other designation given to the material received, other than the donors accession number (23) or collecting number (3). First letter uppercase. Multiple names are separated by a semicolon without space.
     *
     * @return germplasmName
     **/
    @Schema(example = "A0000003", required = true, description = "Name of the germplasm. It can be the preferred name and does not have to be unique.  MCPD (v2.1) (ACCENAME) 11. Either a registered or other designation given to the material received, other than the donors accession number (23) or collecting number (3). First letter uppercase. Multiple names are separated by a semicolon without space.")
    public String getGermplasmName() {
        return germplasmName;
    }

    public void setGermplasmName(String germplasmName) {
        this.germplasmName = germplasmName;
    }

    public GermplasmNewRequest germplasmOrigin(List<GermplasmGermplasmOrigin> germplasmOrigin) {
        this.germplasmOrigin = germplasmOrigin;
        return this;
    }

    public GermplasmNewRequest addGermplasmOriginItem(GermplasmGermplasmOrigin germplasmOriginItem) {
        if (this.germplasmOrigin == null) {
            this.germplasmOrigin = new ArrayList<GermplasmGermplasmOrigin>();
        }
        this.germplasmOrigin.add(germplasmOriginItem);
        return this;
    }

    /**
     * Information for material (orchard, natural sites, ...). Geographic identification of the plants from which seeds or cutting have been taken to produce that germplasm.
     *
     * @return germplasmOrigin
     **/
    @Schema(description = "Information for material (orchard, natural sites, ...). Geographic identification of the plants from which seeds or cutting have been taken to produce that germplasm.")
    public List<GermplasmGermplasmOrigin> getGermplasmOrigin() {
        return germplasmOrigin;
    }

    public void setGermplasmOrigin(List<GermplasmGermplasmOrigin> germplasmOrigin) {
        this.germplasmOrigin = germplasmOrigin;
    }

    public GermplasmNewRequest germplasmPUI(String germplasmPUI) {
        this.germplasmPUI = germplasmPUI;
        return this;
    }

    /**
     * The Permanent Unique Identifier which represents a germplasm  MIAPPE V1.1 (DM-41) Biological material ID - Code used to identify the biological material in the data file. Should be unique within the Investigation. Can correspond to experimental plant ID, seed lot ID, etc This material identification is different from a BiosampleID which corresponds to Observation Unit or Samples sections below.  MIAPPE V1.1 (DM-51) Material source DOI - Digital Object Identifier (DOI) of the material source  MCPD (v2.1) (PUID) 0. Any persistent, unique identifier assigned to the accession so it can be unambiguously referenced at the global level and the information associated with it harvested through automated means. Report one PUID for each accession. The Secretariat of the International Treaty on Plant Genetic Resources for Food and Agriculture (PGRFA) is facilitating the assignment of a persistent unique identifier (PUID), in the form of a DOI, to PGRFA at the accession level. Genebanks not applying a true PUID to their accessions should use, and request recipients to use, the concatenation of INSTCODE, ACCENUMB, and GENUS as a globally unique identifier similar in most respects to the PUID whenever they exchange information on accessions with third parties.
     *
     * @return germplasmPUI
     **/
    @Schema(example = "http://pui.per/accession/A0000003", required = true, description = "The Permanent Unique Identifier which represents a germplasm  MIAPPE V1.1 (DM-41) Biological material ID - Code used to identify the biological material in the data file. Should be unique within the Investigation. Can correspond to experimental plant ID, seed lot ID, etc This material identification is different from a BiosampleID which corresponds to Observation Unit or Samples sections below.  MIAPPE V1.1 (DM-51) Material source DOI - Digital Object Identifier (DOI) of the material source  MCPD (v2.1) (PUID) 0. Any persistent, unique identifier assigned to the accession so it can be unambiguously referenced at the global level and the information associated with it harvested through automated means. Report one PUID for each accession. The Secretariat of the International Treaty on Plant Genetic Resources for Food and Agriculture (PGRFA) is facilitating the assignment of a persistent unique identifier (PUID), in the form of a DOI, to PGRFA at the accession level. Genebanks not applying a true PUID to their accessions should use, and request recipients to use, the concatenation of INSTCODE, ACCENUMB, and GENUS as a globally unique identifier similar in most respects to the PUID whenever they exchange information on accessions with third parties.")
    public String getGermplasmPUI() {
        return germplasmPUI;
    }

    public void setGermplasmPUI(String germplasmPUI) {
        this.germplasmPUI = germplasmPUI;
    }

    public GermplasmNewRequest germplasmPreprocessing(String germplasmPreprocessing) {
        this.germplasmPreprocessing = germplasmPreprocessing;
        return this;
    }

    /**
     * Description of any process or treatment applied uniformly to the germplasm, prior to the study itself. Can be provided as free text or as an accession number from a suitable controlled vocabulary.
     *
     * @return germplasmPreprocessing
     **/
    @Schema(example = "EO:0007210; transplanted from study 2351 observation unit ID: pot:894", description = "Description of any process or treatment applied uniformly to the germplasm, prior to the study itself. Can be provided as free text or as an accession number from a suitable controlled vocabulary.")
    public String getGermplasmPreprocessing() {
        return germplasmPreprocessing;
    }

    public void setGermplasmPreprocessing(String germplasmPreprocessing) {
        this.germplasmPreprocessing = germplasmPreprocessing;
    }

    public GermplasmNewRequest instituteCode(String instituteCode) {
        this.instituteCode = instituteCode;
        return this;
    }

    /**
     * The code for the institute that maintains the material.   MCPD (v2.1) (INSTCODE) 1. FAO WIEWS code of the institute where the accession is maintained. The codes consist of the 3-letter ISO 3166 country code of the country where the institute is located plus a number (e.g. PER001). The current set of institute codes is available from http://www.fao.org/wiews. For those institutes not yet having an FAO Code, or for those with \&quot;obsolete\&quot; codes, see \&quot;Common formatting rules (v)\&quot;.
     *
     * @return instituteCode
     **/
    @Schema(example = "PER001", description = "The code for the institute that maintains the material.   MCPD (v2.1) (INSTCODE) 1. FAO WIEWS code of the institute where the accession is maintained. The codes consist of the 3-letter ISO 3166 country code of the country where the institute is located plus a number (e.g. PER001). The current set of institute codes is available from http://www.fao.org/wiews. For those institutes not yet having an FAO Code, or for those with \"obsolete\" codes, see \"Common formatting rules (v)\".")
    public String getInstituteCode() {
        return instituteCode;
    }

    public void setInstituteCode(String instituteCode) {
        this.instituteCode = instituteCode;
    }

    public GermplasmNewRequest instituteName(String instituteName) {
        this.instituteName = instituteName;
        return this;
    }

    /**
     * The name of the institute that maintains the material
     *
     * @return instituteName
     **/
    @Schema(example = "The BrAPI Institute", description = "The name of the institute that maintains the material")
    public String getInstituteName() {
        return instituteName;
    }

    public void setInstituteName(String instituteName) {
        this.instituteName = instituteName;
    }

    public GermplasmNewRequest pedigree(String pedigree) {
        this.pedigree = pedigree;
        return this;
    }

    /**
     * The cross name and optional selection history.  MCPD (v2.1) (ANCEST) 20. Information about either pedigree or other description of ancestral information (e.g. parent variety in case of mutant or selection). For example a pedigree &#x27;Hanna/7*Atlas//Turk/8*Atlas&#x27; or a description &#x27;mutation found in Hanna&#x27;, &#x27;selection from Irene&#x27; or &#x27;cross involving amongst others Hanna and Irene&#x27;.
     *
     * @return pedigree
     **/
    @Schema(example = "A0000001/A0000002", description = "The cross name and optional selection history.  MCPD (v2.1) (ANCEST) 20. Information about either pedigree or other description of ancestral information (e.g. parent variety in case of mutant or selection). For example a pedigree 'Hanna/7*Atlas//Turk/8*Atlas' or a description 'mutation found in Hanna', 'selection from Irene' or 'cross involving amongst others Hanna and Irene'.")
    public String getPedigree() {
        return pedigree;
    }

    public void setPedigree(String pedigree) {
        this.pedigree = pedigree;
    }

    public GermplasmNewRequest seedSource(String seedSource) {
        this.seedSource = seedSource;
        return this;
    }

    /**
     * An identifier for the source of the biological material &lt;br/&gt;MIAPPE V1.1 (DM-50) Material source ID (Holding institute/stock centre, accession) - An identifier for the source of the biological material, in the form of a key-value pair comprising the name/identifier of the repository from which the material was sourced plus the accession number of the repository for that material. Where an accession number has not been assigned, but the material has been derived from the crossing of known accessions, the material can be defined as follows: \&quot;mother_accession X father_accession\&quot;, or, if father is unknown, as \&quot;mother_accession X UNKNOWN\&quot;. For in situ material, the region of provenance may be used when an accession is not available.
     *
     * @return seedSource
     **/
    @Schema(example = "INRA:095115_inra", description = "An identifier for the source of the biological material <br/>MIAPPE V1.1 (DM-50) Material source ID (Holding institute/stock centre, accession) - An identifier for the source of the biological material, in the form of a key-value pair comprising the name/identifier of the repository from which the material was sourced plus the accession number of the repository for that material. Where an accession number has not been assigned, but the material has been derived from the crossing of known accessions, the material can be defined as follows: \"mother_accession X father_accession\", or, if father is unknown, as \"mother_accession X UNKNOWN\". For in situ material, the region of provenance may be used when an accession is not available.")
    public String getSeedSource() {
        return seedSource;
    }

    public void setSeedSource(String seedSource) {
        this.seedSource = seedSource;
    }

    public GermplasmNewRequest seedSourceDescription(String seedSourceDescription) {
        this.seedSourceDescription = seedSourceDescription;
        return this;
    }

    /**
     * Description of the material source  MIAPPE V1.1 (DM-56) Material source description - Description of the material source
     *
     * @return seedSourceDescription
     **/
    @Schema(example = "Branches were collected from a 10-year-old tree growing in a progeny trial established in a loamy brown earth soil.", description = "Description of the material source  MIAPPE V1.1 (DM-56) Material source description - Description of the material source")
    public String getSeedSourceDescription() {
        return seedSourceDescription;
    }

    public void setSeedSourceDescription(String seedSourceDescription) {
        this.seedSourceDescription = seedSourceDescription;
    }

    public GermplasmNewRequest species(String species) {
        this.species = species;
        return this;
    }

    /**
     * Specific epithet portion of the scientific name in lowercase letters.  MCPD (v2.1) (SPECIES) 6. Specific epithet portion of the scientific name in lowercase letters. Only the following abbreviation is allowed: \&quot;sp.\&quot;   MIAPPE V1.1 (DM-44) Species - Species name (formally: specific epithet) for the organism under study, according to standard scientific nomenclature.
     *
     * @return species
     **/
    @Schema(example = "fructus", description = "Specific epithet portion of the scientific name in lowercase letters.  MCPD (v2.1) (SPECIES) 6. Specific epithet portion of the scientific name in lowercase letters. Only the following abbreviation is allowed: \"sp.\"   MIAPPE V1.1 (DM-44) Species - Species name (formally: specific epithet) for the organism under study, according to standard scientific nomenclature.")
    public String getSpecies() {
        return species;
    }

    public void setSpecies(String species) {
        this.species = species;
    }

    public GermplasmNewRequest speciesAuthority(String speciesAuthority) {
        this.speciesAuthority = speciesAuthority;
        return this;
    }

    /**
     * The authority organization responsible for tracking and maintaining the species name   MCPD (v2.1) (SPAUTHOR) 7. Provide the authority for the species name.
     *
     * @return speciesAuthority
     **/
    @Schema(example = "Smith, 1822", description = "The authority organization responsible for tracking and maintaining the species name   MCPD (v2.1) (SPAUTHOR) 7. Provide the authority for the species name.")
    public String getSpeciesAuthority() {
        return speciesAuthority;
    }

    public void setSpeciesAuthority(String speciesAuthority) {
        this.speciesAuthority = speciesAuthority;
    }

    public GermplasmNewRequest storageTypes(List<GermplasmStorageTypes> storageTypes) {
        this.storageTypes = storageTypes;
        return this;
    }

    public GermplasmNewRequest addStorageTypesItem(GermplasmStorageTypes storageTypesItem) {
        if (this.storageTypes == null) {
            this.storageTypes = new ArrayList<GermplasmStorageTypes>();
        }
        this.storageTypes.add(storageTypesItem);
        return this;
    }

    /**
     * The type of storage this germplasm is kept in at a genebank.
     *
     * @return storageTypes
     **/
    @Schema(example = "[{\"code\":\"20\",\"description\":\"Field collection\"},{\"code\":\"11\",\"description\":\"Short term\"}]", description = "The type of storage this germplasm is kept in at a genebank.")
    public List<GermplasmStorageTypes> getStorageTypes() {
        return storageTypes;
    }

    public void setStorageTypes(List<GermplasmStorageTypes> storageTypes) {
        this.storageTypes = storageTypes;
    }

    public GermplasmNewRequest subtaxa(String subtaxa) {
        this.subtaxa = subtaxa;
        return this;
    }

    /**
     * Subtaxon can be used to store any additional taxonomic identifier.  MCPD (v2.1) (SUBTAXA) 8. Subtaxon can be used to store any additional taxonomic identifier. The following abbreviations are allowed: \&quot;subsp.\&quot; (for subspecies); \&quot;convar.\&quot; (for convariety); \&quot;var.\&quot; (for variety); \&quot;f.\&quot; (for form); \&quot;Group\&quot; (for \&quot;cultivar group\&quot;).  MIAPPE V1.1 (DM-44) Infraspecific name - Name of any subtaxa level, including variety, crossing name, etc. It can be used to store any additional taxonomic identifier. Either free text description or key-value pair list format (the key is the name of the rank and the value is the value of  the rank). Ranks can be among the following terms: subspecies, cultivar, variety, subvariety, convariety, group, subgroup, hybrid, line, form, subform. For MCPD compliance, the following abbreviations are allowed: subsp. (subspecies); convar. (convariety); var. (variety); f. (form); Group (cultivar group).
     *
     * @return subtaxa
     **/
    @Schema(example = "Aspergillus fructus A", description = "Subtaxon can be used to store any additional taxonomic identifier.  MCPD (v2.1) (SUBTAXA) 8. Subtaxon can be used to store any additional taxonomic identifier. The following abbreviations are allowed: \"subsp.\" (for subspecies); \"convar.\" (for convariety); \"var.\" (for variety); \"f.\" (for form); \"Group\" (for \"cultivar group\").  MIAPPE V1.1 (DM-44) Infraspecific name - Name of any subtaxa level, including variety, crossing name, etc. It can be used to store any additional taxonomic identifier. Either free text description or key-value pair list format (the key is the name of the rank and the value is the value of  the rank). Ranks can be among the following terms: subspecies, cultivar, variety, subvariety, convariety, group, subgroup, hybrid, line, form, subform. For MCPD compliance, the following abbreviations are allowed: subsp. (subspecies); convar. (convariety); var. (variety); f. (form); Group (cultivar group).")
    public String getSubtaxa() {
        return subtaxa;
    }

    public void setSubtaxa(String subtaxa) {
        this.subtaxa = subtaxa;
    }

    public GermplasmNewRequest subtaxaAuthority(String subtaxaAuthority) {
        this.subtaxaAuthority = subtaxaAuthority;
        return this;
    }

    /**
     * The authority organization responsible for tracking and maintaining the subtaxon information  MCPD (v2.1) (SUBTAUTHOR) 9. Provide the subtaxon authority at the most detailed taxonomic level.
     *
     * @return subtaxaAuthority
     **/
    @Schema(example = "Smith, 1822", description = "The authority organization responsible for tracking and maintaining the subtaxon information  MCPD (v2.1) (SUBTAUTHOR) 9. Provide the subtaxon authority at the most detailed taxonomic level.")
    public String getSubtaxaAuthority() {
        return subtaxaAuthority;
    }

    public void setSubtaxaAuthority(String subtaxaAuthority) {
        this.subtaxaAuthority = subtaxaAuthority;
    }

    public GermplasmNewRequest synonyms(List<GermplasmSynonyms> synonyms) {
        this.synonyms = synonyms;
        return this;
    }

    public GermplasmNewRequest addSynonymsItem(GermplasmSynonyms synonymsItem) {
        if (this.synonyms == null) {
            this.synonyms = new ArrayList<GermplasmSynonyms>();
        }
        this.synonyms.add(synonymsItem);
        return this;
    }

    /**
     * List of alternative names or IDs used to reference this germplasm  MCPD (v2.1) (OTHERNUMB) 24. Any other identifiers known to exist in other collections for this accession. Use the following format: INSTCODE:ACCENUMB;INSTCODE:identifier;INSTCODE and identifier are separated by a colon without space. Pairs of INSTCODE and identifier are separated by a semicolon without space. When the institute is not known, the identifier should be preceded by a colon.
     *
     * @return synonyms
     **/
    @Schema(description = "List of alternative names or IDs used to reference this germplasm  MCPD (v2.1) (OTHERNUMB) 24. Any other identifiers known to exist in other collections for this accession. Use the following format: INSTCODE:ACCENUMB;INSTCODE:identifier;INSTCODE and identifier are separated by a colon without space. Pairs of INSTCODE and identifier are separated by a semicolon without space. When the institute is not known, the identifier should be preceded by a colon.")
    public List<GermplasmSynonyms> getSynonyms() {
        return synonyms;
    }

    public void setSynonyms(List<GermplasmSynonyms> synonyms) {
        this.synonyms = synonyms;
    }

    public GermplasmNewRequest taxonIds(List<GermplasmTaxonIds> taxonIds) {
        this.taxonIds = taxonIds;
        return this;
    }

    public GermplasmNewRequest addTaxonIdsItem(GermplasmTaxonIds taxonIdsItem) {
        if (this.taxonIds == null) {
            this.taxonIds = new ArrayList<GermplasmTaxonIds>();
        }
        this.taxonIds.add(taxonIdsItem);
        return this;
    }

    /**
     * The list of IDs for this SPECIES from different sources. If present, NCBI Taxon should be always listed as \&quot;ncbiTaxon\&quot; preferably with a purl. The rank of this ID should be species.  MIAPPE V1.1 (DM-42) Organism - An identifier for the organism at the species level. Use of the NCBI taxon ID is recommended.
     *
     * @return taxonIds
     **/
    @Schema(description = "The list of IDs for this SPECIES from different sources. If present, NCBI Taxon should be always listed as \"ncbiTaxon\" preferably with a purl. The rank of this ID should be species.  MIAPPE V1.1 (DM-42) Organism - An identifier for the organism at the species level. Use of the NCBI taxon ID is recommended.")
    public List<GermplasmTaxonIds> getTaxonIds() {
        return taxonIds;
    }

    public void setTaxonIds(List<GermplasmTaxonIds> taxonIds) {
        this.taxonIds = taxonIds;
    }


    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        GermplasmNewRequest germplasmNewRequest = (GermplasmNewRequest) o;
        return Objects.equals(this.accessionNumber, germplasmNewRequest.accessionNumber) &&
                Objects.equals(this.acquisitionDate, germplasmNewRequest.acquisitionDate) &&
                Objects.equals(this.additionalInfo, germplasmNewRequest.additionalInfo) &&
                Objects.equals(this.biologicalStatusOfAccessionCode, germplasmNewRequest.biologicalStatusOfAccessionCode) &&
                Objects.equals(this.biologicalStatusOfAccessionDescription, germplasmNewRequest.biologicalStatusOfAccessionDescription) &&
                Objects.equals(this.breedingMethodDbId, germplasmNewRequest.breedingMethodDbId) &&
                Objects.equals(this.breedingMethodName, germplasmNewRequest.breedingMethodName) &&
                Objects.equals(this.collection, germplasmNewRequest.collection) &&
                Objects.equals(this.commonCropName, germplasmNewRequest.commonCropName) &&
                Objects.equals(this.countryOfOriginCode, germplasmNewRequest.countryOfOriginCode) &&
                Objects.equals(this.defaultDisplayName, germplasmNewRequest.defaultDisplayName) &&
                Objects.equals(this.documentationURL, germplasmNewRequest.documentationURL) &&
                Objects.equals(this.donors, germplasmNewRequest.donors) &&
                Objects.equals(this.externalReferences, germplasmNewRequest.externalReferences) &&
                Objects.equals(this.genus, germplasmNewRequest.genus) &&
                Objects.equals(this.germplasmName, germplasmNewRequest.germplasmName) &&
                Objects.equals(this.germplasmOrigin, germplasmNewRequest.germplasmOrigin) &&
                Objects.equals(this.germplasmPUI, germplasmNewRequest.germplasmPUI) &&
                Objects.equals(this.germplasmPreprocessing, germplasmNewRequest.germplasmPreprocessing) &&
                Objects.equals(this.instituteCode, germplasmNewRequest.instituteCode) &&
                Objects.equals(this.instituteName, germplasmNewRequest.instituteName) &&
                Objects.equals(this.pedigree, germplasmNewRequest.pedigree) &&
                Objects.equals(this.seedSource, germplasmNewRequest.seedSource) &&
                Objects.equals(this.seedSourceDescription, germplasmNewRequest.seedSourceDescription) &&
                Objects.equals(this.species, germplasmNewRequest.species) &&
                Objects.equals(this.speciesAuthority, germplasmNewRequest.speciesAuthority) &&
                Objects.equals(this.storageTypes, germplasmNewRequest.storageTypes) &&
                Objects.equals(this.subtaxa, germplasmNewRequest.subtaxa) &&
                Objects.equals(this.subtaxaAuthority, germplasmNewRequest.subtaxaAuthority) &&
                Objects.equals(this.synonyms, germplasmNewRequest.synonyms) &&
                Objects.equals(this.taxonIds, germplasmNewRequest.taxonIds);
    }

    @Override
    public int hashCode() {
        return Objects.hash(accessionNumber, acquisitionDate, additionalInfo, biologicalStatusOfAccessionCode, biologicalStatusOfAccessionDescription, breedingMethodDbId, breedingMethodName, collection, commonCropName, countryOfOriginCode, defaultDisplayName, documentationURL, donors, externalReferences, genus, germplasmName, germplasmOrigin, germplasmPUI, germplasmPreprocessing, instituteCode, instituteName, pedigree, seedSource, seedSourceDescription, species, speciesAuthority, storageTypes, subtaxa, subtaxaAuthority, synonyms, taxonIds);
    }


    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("class GermplasmNewRequest {\n");

        sb.append("    accessionNumber: ").append(toIndentedString(accessionNumber)).append("\n");
        sb.append("    acquisitionDate: ").append(toIndentedString(acquisitionDate)).append("\n");
        sb.append("    additionalInfo: ").append(toIndentedString(additionalInfo)).append("\n");
        sb.append("    biologicalStatusOfAccessionCode: ").append(toIndentedString(biologicalStatusOfAccessionCode)).append("\n");
        sb.append("    biologicalStatusOfAccessionDescription: ").append(toIndentedString(biologicalStatusOfAccessionDescription)).append("\n");
        sb.append("    breedingMethodDbId: ").append(toIndentedString(breedingMethodDbId)).append("\n");
        sb.append("    breedingMethodName: ").append(toIndentedString(breedingMethodName)).append("\n");
        sb.append("    collection: ").append(toIndentedString(collection)).append("\n");
        sb.append("    commonCropName: ").append(toIndentedString(commonCropName)).append("\n");
        sb.append("    countryOfOriginCode: ").append(toIndentedString(countryOfOriginCode)).append("\n");
        sb.append("    defaultDisplayName: ").append(toIndentedString(defaultDisplayName)).append("\n");
        sb.append("    documentationURL: ").append(toIndentedString(documentationURL)).append("\n");
        sb.append("    donors: ").append(toIndentedString(donors)).append("\n");
        sb.append("    externalReferences: ").append(toIndentedString(externalReferences)).append("\n");
        sb.append("    genus: ").append(toIndentedString(genus)).append("\n");
        sb.append("    germplasmName: ").append(toIndentedString(germplasmName)).append("\n");
        sb.append("    germplasmOrigin: ").append(toIndentedString(germplasmOrigin)).append("\n");
        sb.append("    germplasmPUI: ").append(toIndentedString(germplasmPUI)).append("\n");
        sb.append("    germplasmPreprocessing: ").append(toIndentedString(germplasmPreprocessing)).append("\n");
        sb.append("    instituteCode: ").append(toIndentedString(instituteCode)).append("\n");
        sb.append("    instituteName: ").append(toIndentedString(instituteName)).append("\n");
        sb.append("    pedigree: ").append(toIndentedString(pedigree)).append("\n");
        sb.append("    seedSource: ").append(toIndentedString(seedSource)).append("\n");
        sb.append("    seedSourceDescription: ").append(toIndentedString(seedSourceDescription)).append("\n");
        sb.append("    species: ").append(toIndentedString(species)).append("\n");
        sb.append("    speciesAuthority: ").append(toIndentedString(speciesAuthority)).append("\n");
        sb.append("    storageTypes: ").append(toIndentedString(storageTypes)).append("\n");
        sb.append("    subtaxa: ").append(toIndentedString(subtaxa)).append("\n");
        sb.append("    subtaxaAuthority: ").append(toIndentedString(subtaxaAuthority)).append("\n");
        sb.append("    synonyms: ").append(toIndentedString(synonyms)).append("\n");
        sb.append("    taxonIds: ").append(toIndentedString(taxonIds)).append("\n");
        sb.append("}");
        return sb.toString();
    }

    /**
     * Convert the given object to string with each line indented by 4 spaces
     * (except the first line).
     */
    private String toIndentedString(Object o) {
        if (o == null) {
            return "null";
        }
        return o.toString().replace("\n", "\n    ");
    }

}
