/*
 * BrAPI-Genotyping
 * The Breeding API (BrAPI) is a Standardized REST ful Web Service API Specification for communicating Plant Breeding Data. BrAPI allows for easy data sharing between databases and tools involved in plant breeding. <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">General Reference Documentation</h2> <div class=\"gen-info-link\"><a href=\"https://wiki.brapi.org/index.php/RESTfulness\">URL Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://wiki.brapi.org/index.php/Response_Structure\">Response Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://wiki.brapi.org/index.php/Dates_and_Times\">Date/Time Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://wiki.brapi.org/index.php/Location_Coordinates\">Location Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://wiki.brapi.org/index.php/Error_Handling\">Error Handling</a></div> <div class=\"gen-info-link\"><a href=\"https://wiki.brapi.org/index.php/Search_Services\">Search Services</a></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Core</h2> <div class=\"brapi-section-description\">The BrAPI Core module contains high level entities used for organization and management. This includes Programs, Trials, Studies, Locations, People, and Lists</div> <div class=\"version-number\">V2.1</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/brapi-V2.1/Specification/BrAPI-Core\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Core/2.1\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapicore21.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Phenotyping</h2> <div class=\"brapi-section-description\">The BrAPI Phenotyping module contains entities related to phenotypic observations. This includes Observation Units, Observations, Observation Variables, Traits, Scales, Methods, and Images</div> <div class=\"version-number\">V2.1</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/brapi-V2.1/Specification/BrAPI-Phenotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Phenotyping/2.1\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapiphenotyping21.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"current-brapi-section brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Genotyping</h2> <div class=\"brapi-section-description\">The BrAPI Genotyping module contains entities related to genotyping analysis. This includes Samples, Markers, Variant Sets, Variants, Call Sets, Calls, References, Reads, and Vendor Orders</div> <div class=\"version-number\">V2.1</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/brapi-V2.1/Specification/BrAPI-Genotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Genotyping/2.1\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigenotyping21.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Germplasm</h2> <div class=\"brapi-section-description\">The BrAPI Germplasm module contains entities related to germplasm management. This includes Germplasm, Germplasm Attributes, Seed Lots, Crosses, Pedigree, and Progeny</div> <div class=\"version-number\">V2.1</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/brapi-V2.1/Specification/BrAPI-Germplasm\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Germplasm/2.1\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigermplasm21.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <style> .link-btn{ float: left;  margin: 2px 10px 0 0;  padding: 0 5px;  border-radius: 5px;  background-color: #ddd; } .stop-float{   clear: both; } .version-number{   float: left;    margin: 5px 10px 0 5px; } .brapi-section-title{   margin: 0 10px 0 0;   font-size: 20px; } .current-brapi-section{   font-weight: bolder;   border-radius: 5px;    background-color: #ddd; } .brapi-section{   padding: 5px 5px;  } .brapi-section-description{   margin: 5px 0 0 5px; } </style>
 *
 * OpenAPI spec version: 2.1
 *
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 */

package org.brapi.model.v21.genotype;

import com.google.gson.annotations.SerializedName;
import io.swagger.v3.oas.annotations.media.Schema;

import java.util.*;

/**
 * A &#x60;Call&#x60; represents the determination of genotype with respect to a particular &#x60;Variant&#x60;.  It may include associated information such as quality and phasing. For example, a call might assign a probability of 0.32 to the occurrence of a SNP named RS_1234 in a call set with the name NA_12345.
 */
@Schema(description = "A `Call` represents the determination of genotype with respect to a particular `Variant`.  It may include associated information such as quality and phasing. For example, a call might assign a probability of 0.32 to the occurrence of a SNP named RS_1234 in a call set with the name NA_12345.")
@javax.annotation.Generated(value = "io.swagger.codegen.v3.generators.java.JavaClientCodegen", date = "2022-10-25T14:54:00.515Z[GMT]")
public class Call {
    @SerializedName("additionalInfo")
    private Map<String, String> additionalInfo = null;

    @SerializedName("callSetDbId")
    private String callSetDbId = null;

    @SerializedName("callSetName")
    private String callSetName = null;

    @SerializedName("genotype")
    private ListValue genotype = null;

    @SerializedName("genotypeMetadata")
    private List<CallGenotypeMetadata> genotypeMetadata = null;

    @SerializedName("genotypeValue")
    private String genotypeValue = null;

    @SerializedName("genotype_likelihood")
    private List<Double> genotypeLikelihood = null;

    @SerializedName("phaseSet")
    private String phaseSet = null;

    @SerializedName("variantDbId")
    private String variantDbId = null;

    @SerializedName("variantName")
    private String variantName = null;

    @SerializedName("variantSetDbId")
    private String variantSetDbId = null;

    @SerializedName("variantSetName")
    private String variantSetName = null;

    public Call additionalInfo(Map<String, String> additionalInfo) {
        this.additionalInfo = additionalInfo;
        return this;
    }

    public Call putAdditionalInfoItem(String key, String additionalInfoItem) {
        if (this.additionalInfo == null) {
            this.additionalInfo = new HashMap<String, String>();
        }
        this.additionalInfo.put(key, additionalInfoItem);
        return this;
    }

    /**
     * A free space containing any additional information related to a particular object. A data source may provide any JSON object, unrestriced by the BrAPI specification.
     *
     * @return additionalInfo
     **/
    @Schema(description = "A free space containing any additional information related to a particular object. A data source may provide any JSON object, unrestriced by the BrAPI specification.")
    public Map<String, String> getAdditionalInfo() {
        return additionalInfo;
    }

    public void setAdditionalInfo(Map<String, String> additionalInfo) {
        this.additionalInfo = additionalInfo;
    }

    public Call callSetDbId(String callSetDbId) {
        this.callSetDbId = callSetDbId;
        return this;
    }

    /**
     * The ID of the call set this variant call belongs to.  If this field is not present, the ordering of the call sets from a &#x60;SearchCallSetsRequest&#x60; over this &#x60;VariantSet&#x60; is guaranteed to match the ordering of the calls on this &#x60;Variant&#x60;. The number of results will also be the same.
     *
     * @return callSetDbId
     **/
    @Schema(example = "16466f55", description = "The ID of the call set this variant call belongs to.  If this field is not present, the ordering of the call sets from a `SearchCallSetsRequest` over this `VariantSet` is guaranteed to match the ordering of the calls on this `Variant`. The number of results will also be the same.")
    public String getCallSetDbId() {
        return callSetDbId;
    }

    public void setCallSetDbId(String callSetDbId) {
        this.callSetDbId = callSetDbId;
    }

    public Call callSetName(String callSetName) {
        this.callSetName = callSetName;
        return this;
    }

    /**
     * The name of the call set this variant call belongs to. If this field is not present, the ordering of the call sets from a &#x60;SearchCallSetsRequest&#x60; over this &#x60;VariantSet&#x60; is guaranteed to match the ordering of the calls on this &#x60;Variant&#x60;. The number of results will also be the same.
     *
     * @return callSetName
     **/
    @Schema(example = "Sample_123_DNA_Run_456", description = "The name of the call set this variant call belongs to. If this field is not present, the ordering of the call sets from a `SearchCallSetsRequest` over this `VariantSet` is guaranteed to match the ordering of the calls on this `Variant`. The number of results will also be the same.")
    public String getCallSetName() {
        return callSetName;
    }

    public void setCallSetName(String callSetName) {
        this.callSetName = callSetName;
    }

    public Call genotype(ListValue genotype) {
        this.genotype = genotype;
        return this;
    }

    /**
     * Get genotype
     *
     * @return genotype
     **/
    @Schema(description = "")
    public ListValue getGenotype() {
        return genotype;
    }

    public void setGenotype(ListValue genotype) {
        this.genotype = genotype;
    }

    public Call genotypeMetadata(List<CallGenotypeMetadata> genotypeMetadata) {
        this.genotypeMetadata = genotypeMetadata;
        return this;
    }

    public Call addGenotypeMetadataItem(CallGenotypeMetadata genotypeMetadataItem) {
        if (this.genotypeMetadata == null) {
            this.genotypeMetadata = new ArrayList<CallGenotypeMetadata>();
        }
        this.genotypeMetadata.add(genotypeMetadataItem);
        return this;
    }

    /**
     * Genotype Metadata are additional layers of metadata associated with each genotype.
     *
     * @return genotypeMetadata
     **/
    @Schema(description = "Genotype Metadata are additional layers of metadata associated with each genotype.")
    public List<CallGenotypeMetadata> getGenotypeMetadata() {
        return genotypeMetadata;
    }

    public void setGenotypeMetadata(List<CallGenotypeMetadata> genotypeMetadata) {
        this.genotypeMetadata = genotypeMetadata;
    }

    public Call genotypeValue(String genotypeValue) {
        this.genotypeValue = genotypeValue;
        return this;
    }

    /**
     * The value of this genotype call
     *
     * @return genotypeValue
     **/
    @Schema(example = "1/1", description = "The value of this genotype call")
    public String getGenotypeValue() {
        return genotypeValue;
    }

    public void setGenotypeValue(String genotypeValue) {
        this.genotypeValue = genotypeValue;
    }

    public Call genotypeLikelihood(List<Double> genotypeLikelihood) {
        this.genotypeLikelihood = genotypeLikelihood;
        return this;
    }

    public Call addGenotypeLikelihoodItem(Double genotypeLikelihoodItem) {
        if (this.genotypeLikelihood == null) {
            this.genotypeLikelihood = new ArrayList<Double>();
        }
        this.genotypeLikelihood.add(genotypeLikelihoodItem);
        return this;
    }

    /**
     * **Deprecated in v2.1** Please use &#x60;genotypeMetadata&#x60;. Github issue number #491              &lt;br&gt;The genotype likelihood for this variant call. Each array entry represents how likely a specific genotype is for this call as log10(P(data | genotype)), analogous to the GL tag in the VCF spec. The value ordering is defined by the GL tag in the VCF spec.
     *
     * @return genotypeLikelihood
     **/
    @Schema(example = "[1]", description = "**Deprecated in v2.1** Please use `genotypeMetadata`. Github issue number #491              <br>The genotype likelihood for this variant call. Each array entry represents how likely a specific genotype is for this call as log10(P(data | genotype)), analogous to the GL tag in the VCF spec. The value ordering is defined by the GL tag in the VCF spec.")
    public List<Double> getGenotypeLikelihood() {
        return genotypeLikelihood;
    }

    public void setGenotypeLikelihood(List<Double> genotypeLikelihood) {
        this.genotypeLikelihood = genotypeLikelihood;
    }

    public Call phaseSet(String phaseSet) {
        this.phaseSet = phaseSet;
        return this;
    }

    /**
     * If this field is populated, this variant call&#x27;s genotype ordering implies the phase of the bases and  is consistent with any other variant calls on the same contig which have the same phase set string.
     *
     * @return phaseSet
     **/
    @Schema(example = "6410afc5", description = "If this field is populated, this variant call's genotype ordering implies the phase of the bases and  is consistent with any other variant calls on the same contig which have the same phase set string.")
    public String getPhaseSet() {
        return phaseSet;
    }

    public void setPhaseSet(String phaseSet) {
        this.phaseSet = phaseSet;
    }

    public Call variantDbId(String variantDbId) {
        this.variantDbId = variantDbId;
        return this;
    }

    /**
     * The ID of the variant this call belongs to.
     *
     * @return variantDbId
     **/
    @Schema(example = "538c8ecf", description = "The ID of the variant this call belongs to.")
    public String getVariantDbId() {
        return variantDbId;
    }

    public void setVariantDbId(String variantDbId) {
        this.variantDbId = variantDbId;
    }

    public Call variantName(String variantName) {
        this.variantName = variantName;
        return this;
    }

    /**
     * The name of the variant this call belongs to.
     *
     * @return variantName
     **/
    @Schema(example = "Marker A", description = "The name of the variant this call belongs to.")
    public String getVariantName() {
        return variantName;
    }

    public void setVariantName(String variantName) {
        this.variantName = variantName;
    }

    public Call variantSetDbId(String variantSetDbId) {
        this.variantSetDbId = variantSetDbId;
        return this;
    }

    /**
     * The unique identifier for a VariantSet
     *
     * @return variantSetDbId
     **/
    @Schema(example = "87a6ac1e", description = "The unique identifier for a VariantSet")
    public String getVariantSetDbId() {
        return variantSetDbId;
    }

    public void setVariantSetDbId(String variantSetDbId) {
        this.variantSetDbId = variantSetDbId;
    }

    public Call variantSetName(String variantSetName) {
        this.variantSetName = variantSetName;
        return this;
    }

    /**
     * The human readable name for a VariantSet
     *
     * @return variantSetName
     **/
    @Schema(example = "Maize QC DataSet 002334", description = "The human readable name for a VariantSet")
    public String getVariantSetName() {
        return variantSetName;
    }

    public void setVariantSetName(String variantSetName) {
        this.variantSetName = variantSetName;
    }


    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Call call = (Call) o;
        return Objects.equals(this.additionalInfo, call.additionalInfo) &&
                Objects.equals(this.callSetDbId, call.callSetDbId) &&
                Objects.equals(this.callSetName, call.callSetName) &&
                Objects.equals(this.genotype, call.genotype) &&
                Objects.equals(this.genotypeMetadata, call.genotypeMetadata) &&
                Objects.equals(this.genotypeValue, call.genotypeValue) &&
                Objects.equals(this.genotypeLikelihood, call.genotypeLikelihood) &&
                Objects.equals(this.phaseSet, call.phaseSet) &&
                Objects.equals(this.variantDbId, call.variantDbId) &&
                Objects.equals(this.variantName, call.variantName) &&
                Objects.equals(this.variantSetDbId, call.variantSetDbId) &&
                Objects.equals(this.variantSetName, call.variantSetName);
    }

    @Override
    public int hashCode() {
        return Objects.hash(additionalInfo, callSetDbId, callSetName, genotype, genotypeMetadata, genotypeValue, genotypeLikelihood, phaseSet, variantDbId, variantName, variantSetDbId, variantSetName);
    }


    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("class Call {\n");

        sb.append("    additionalInfo: ").append(toIndentedString(additionalInfo)).append("\n");
        sb.append("    callSetDbId: ").append(toIndentedString(callSetDbId)).append("\n");
        sb.append("    callSetName: ").append(toIndentedString(callSetName)).append("\n");
        sb.append("    genotype: ").append(toIndentedString(genotype)).append("\n");
        sb.append("    genotypeMetadata: ").append(toIndentedString(genotypeMetadata)).append("\n");
        sb.append("    genotypeValue: ").append(toIndentedString(genotypeValue)).append("\n");
        sb.append("    genotypeLikelihood: ").append(toIndentedString(genotypeLikelihood)).append("\n");
        sb.append("    phaseSet: ").append(toIndentedString(phaseSet)).append("\n");
        sb.append("    variantDbId: ").append(toIndentedString(variantDbId)).append("\n");
        sb.append("    variantName: ").append(toIndentedString(variantName)).append("\n");
        sb.append("    variantSetDbId: ").append(toIndentedString(variantSetDbId)).append("\n");
        sb.append("    variantSetName: ").append(toIndentedString(variantSetName)).append("\n");
        sb.append("}");
        return sb.toString();
    }

    /**
     * Convert the given object to string with each line indented by 4 spaces
     * (except the first line).
     */
    private String toIndentedString(Object o) {
        if (o == null) {
            return "null";
        }
        return o.toString().replace("\n", "\n    ");
    }

}
