/*
 * BrAPI-Genotyping
 * The Breeding API (BrAPI) is a Standardized REST ful Web Service API Specification for communicating Plant Breeding Data. BrAPI allows for easy data sharing between databases and tools involved in plant breeding. <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">General Reference Documentation</h2> <div class=\"gen-info-link\"><a href=\"https://wiki.brapi.org/index.php/RESTfulness\">URL Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://wiki.brapi.org/index.php/Response_Structure\">Response Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://wiki.brapi.org/index.php/Dates_and_Times\">Date/Time Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://wiki.brapi.org/index.php/Location_Coordinates\">Location Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://wiki.brapi.org/index.php/Error_Handling\">Error Handling</a></div> <div class=\"gen-info-link\"><a href=\"https://wiki.brapi.org/index.php/Search_Services\">Search Services</a></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Core</h2> <div class=\"brapi-section-description\">The BrAPI Core module contains high level entities used for organization and management. This includes Programs, Trials, Studies, Locations, People, and Lists</div> <div class=\"version-number\">V2.1</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/brapi-V2.1/Specification/BrAPI-Core\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Core/2.1\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapicore21.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Phenotyping</h2> <div class=\"brapi-section-description\">The BrAPI Phenotyping module contains entities related to phenotypic observations. This includes Observation Units, Observations, Observation Variables, Traits, Scales, Methods, and Images</div> <div class=\"version-number\">V2.1</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/brapi-V2.1/Specification/BrAPI-Phenotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Phenotyping/2.1\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapiphenotyping21.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"current-brapi-section brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Genotyping</h2> <div class=\"brapi-section-description\">The BrAPI Genotyping module contains entities related to genotyping analysis. This includes Samples, Markers, Variant Sets, Variants, Call Sets, Calls, References, Reads, and Vendor Orders</div> <div class=\"version-number\">V2.1</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/brapi-V2.1/Specification/BrAPI-Genotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Genotyping/2.1\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigenotyping21.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Germplasm</h2> <div class=\"brapi-section-description\">The BrAPI Germplasm module contains entities related to germplasm management. This includes Germplasm, Germplasm Attributes, Seed Lots, Crosses, Pedigree, and Progeny</div> <div class=\"version-number\">V2.1</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/brapi-V2.1/Specification/BrAPI-Germplasm\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Germplasm/2.1\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigermplasm21.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <style> .link-btn{ float: left;  margin: 2px 10px 0 0;  padding: 0 5px;  border-radius: 5px;  background-color: #ddd; } .stop-float{   clear: both; } .version-number{   float: left;    margin: 5px 10px 0 5px; } .brapi-section-title{   margin: 0 10px 0 0;   font-size: 20px; } .current-brapi-section{   font-weight: bolder;   border-radius: 5px;    background-color: #ddd; } .brapi-section{   padding: 5px 5px;  } .brapi-section-description{   margin: 5px 0 0 5px; } </style>
 *
 * OpenAPI spec version: 2.1
 *
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 */

package org.brapi.model.v21.genotype;

import com.google.gson.annotations.SerializedName;
import io.swagger.v3.oas.annotations.media.Schema;

import java.util.*;

/**
 * A &#x60;ReferenceSet&#x60; is a set of &#x60;Reference&#x60; s which typically comprise a reference assembly, such as &#x60;GRCH_38&#x60;. A &#x60;ReferenceSet&#x60; defines a common coordinate space for comparing reference-aligned experimental data.
 */
@Schema(description = "A `ReferenceSet` is a set of `Reference` s which typically comprise a reference assembly, such as `GRCH_38`. A `ReferenceSet` defines a common coordinate space for comparing reference-aligned experimental data.")
@javax.annotation.Generated(value = "io.swagger.codegen.v3.generators.java.JavaClientCodegen", date = "2022-10-25T14:54:00.515Z[GMT]")
public class ReferenceSet {
    @SerializedName("additionalInfo")
    private Map<String, String> additionalInfo = null;

    @SerializedName("assemblyPUI")
    private String assemblyPUI = null;

    @SerializedName("commonCropName")
    private String commonCropName = null;

    @SerializedName("description")
    private String description = null;

    @SerializedName("externalReferences")
    private ExternalReferences externalReferences = null;

    @SerializedName("isDerived")
    private Boolean isDerived = null;

    @SerializedName("md5checksum")
    private String md5checksum = null;

    @SerializedName("referenceSetDbId")
    private String referenceSetDbId = null;

    @SerializedName("referenceSetName")
    private String referenceSetName = null;

    @SerializedName("sourceAccessions")
    private List<String> sourceAccessions = null;

    @SerializedName("sourceGermplasm")
    private List<ReferenceSetSourceGermplasm> sourceGermplasm = null;

    @SerializedName("sourceURI")
    private String sourceURI = null;

    @SerializedName("species")
    private OntologyTerm species = null;

    public ReferenceSet additionalInfo(Map<String, String> additionalInfo) {
        this.additionalInfo = additionalInfo;
        return this;
    }

    public ReferenceSet putAdditionalInfoItem(String key, String additionalInfoItem) {
        if (this.additionalInfo == null) {
            this.additionalInfo = new HashMap<String, String>();
        }
        this.additionalInfo.put(key, additionalInfoItem);
        return this;
    }

    /**
     * A free space containing any additional information related to a particular object. A data source may provide any JSON object, unrestriced by the BrAPI specification.
     *
     * @return additionalInfo
     **/
    @Schema(description = "A free space containing any additional information related to a particular object. A data source may provide any JSON object, unrestriced by the BrAPI specification.")
    public Map<String, String> getAdditionalInfo() {
        return additionalInfo;
    }

    public void setAdditionalInfo(Map<String, String> additionalInfo) {
        this.additionalInfo = additionalInfo;
    }

    public ReferenceSet assemblyPUI(String assemblyPUI) {
        this.assemblyPUI = assemblyPUI;
        return this;
    }

    /**
     * The remaining information is about the source of the sequences Public id of this reference set, such as &#x60;GRCH_37&#x60;.
     *
     * @return assemblyPUI
     **/
    @Schema(example = "doi://10.12345/fake/9876", description = "The remaining information is about the source of the sequences Public id of this reference set, such as `GRCH_37`.")
    public String getAssemblyPUI() {
        return assemblyPUI;
    }

    public void setAssemblyPUI(String assemblyPUI) {
        this.assemblyPUI = assemblyPUI;
    }

    public ReferenceSet commonCropName(String commonCropName) {
        this.commonCropName = commonCropName;
        return this;
    }

    /**
     * Common name for the crop
     *
     * @return commonCropName
     **/
    @Schema(example = "Maize", description = "Common name for the crop")
    public String getCommonCropName() {
        return commonCropName;
    }

    public void setCommonCropName(String commonCropName) {
        this.commonCropName = commonCropName;
    }

    public ReferenceSet description(String description) {
        this.description = description;
        return this;
    }

    /**
     * Optional free text description of this reference set.
     *
     * @return description
     **/
    @Schema(example = "This is an example description for an assembly", description = "Optional free text description of this reference set.")
    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public ReferenceSet externalReferences(ExternalReferences externalReferences) {
        this.externalReferences = externalReferences;
        return this;
    }

    /**
     * Get externalReferences
     *
     * @return externalReferences
     **/
    @Schema(description = "")
    public ExternalReferences getExternalReferences() {
        return externalReferences;
    }

    public void setExternalReferences(ExternalReferences externalReferences) {
        this.externalReferences = externalReferences;
    }

    public ReferenceSet isDerived(Boolean isDerived) {
        this.isDerived = isDerived;
        return this;
    }

    /**
     * A reference set may be derived from a source if it contains additional sequences, or some of the sequences within it are derived (see the definition of &#x60;isDerived&#x60; in &#x60;Reference&#x60;).
     *
     * @return isDerived
     **/
    @Schema(description = "A reference set may be derived from a source if it contains additional sequences, or some of the sequences within it are derived (see the definition of `isDerived` in `Reference`).")
    public Boolean isIsDerived() {
        return isDerived;
    }

    public void setIsDerived(Boolean isDerived) {
        this.isDerived = isDerived;
    }

    public ReferenceSet md5checksum(String md5checksum) {
        this.md5checksum = md5checksum;
        return this;
    }

    /**
     * Order-independent MD5 checksum which identifies this &#x60;ReferenceSet&#x60;.  To compute this checksum, make a list of &#x60;Reference.md5checksum&#x60; for all &#x60;Reference&#x60; s in this set. Then sort that list, and take the MD5 hash of all the strings concatenated together. Express the hash as a lower-case hexadecimal string.
     *
     * @return md5checksum
     **/
    @Schema(example = "c2365e900c81a89cf74d83dab60df146", description = "Order-independent MD5 checksum which identifies this `ReferenceSet`.  To compute this checksum, make a list of `Reference.md5checksum` for all `Reference` s in this set. Then sort that list, and take the MD5 hash of all the strings concatenated together. Express the hash as a lower-case hexadecimal string.")
    public String getMd5checksum() {
        return md5checksum;
    }

    public void setMd5checksum(String md5checksum) {
        this.md5checksum = md5checksum;
    }

    public ReferenceSet referenceSetDbId(String referenceSetDbId) {
        this.referenceSetDbId = referenceSetDbId;
        return this;
    }

    /**
     * The unique identifier for a ReferenceSet
     *
     * @return referenceSetDbId
     **/
    @Schema(example = "c1ecfef1", description = "The unique identifier for a ReferenceSet")
    public String getReferenceSetDbId() {
        return referenceSetDbId;
    }

    public void setReferenceSetDbId(String referenceSetDbId) {
        this.referenceSetDbId = referenceSetDbId;
    }

    public ReferenceSet referenceSetName(String referenceSetName) {
        this.referenceSetName = referenceSetName;
        return this;
    }

    /**
     * The human readable name of a ReferenceSet
     *
     * @return referenceSetName
     **/
    @Schema(example = "The Best Assembly Ever", description = "The human readable name of a ReferenceSet")
    public String getReferenceSetName() {
        return referenceSetName;
    }

    public void setReferenceSetName(String referenceSetName) {
        this.referenceSetName = referenceSetName;
    }

    public ReferenceSet sourceAccessions(List<String> sourceAccessions) {
        this.sourceAccessions = sourceAccessions;
        return this;
    }

    public ReferenceSet addSourceAccessionsItem(String sourceAccessionsItem) {
        if (this.sourceAccessions == null) {
            this.sourceAccessions = new ArrayList<String>();
        }
        this.sourceAccessions.add(sourceAccessionsItem);
        return this;
    }

    /**
     * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally with a version number, e.g. &#x60;NC_000001.11&#x60;.
     *
     * @return sourceAccessions
     **/
    @Schema(example = "[\"A0000002\",\"A0009393\"]", description = "All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally with a version number, e.g. `NC_000001.11`.")
    public List<String> getSourceAccessions() {
        return sourceAccessions;
    }

    public void setSourceAccessions(List<String> sourceAccessions) {
        this.sourceAccessions = sourceAccessions;
    }

    public ReferenceSet sourceGermplasm(List<ReferenceSetSourceGermplasm> sourceGermplasm) {
        this.sourceGermplasm = sourceGermplasm;
        return this;
    }

    public ReferenceSet addSourceGermplasmItem(ReferenceSetSourceGermplasm sourceGermplasmItem) {
        if (this.sourceGermplasm == null) {
            this.sourceGermplasm = new ArrayList<ReferenceSetSourceGermplasm>();
        }
        this.sourceGermplasm.add(sourceGermplasmItem);
        return this;
    }

    /**
     * All known corresponding Germplasm
     *
     * @return sourceGermplasm
     **/
    @Schema(description = "All known corresponding Germplasm")
    public List<ReferenceSetSourceGermplasm> getSourceGermplasm() {
        return sourceGermplasm;
    }

    public void setSourceGermplasm(List<ReferenceSetSourceGermplasm> sourceGermplasm) {
        this.sourceGermplasm = sourceGermplasm;
    }

    public ReferenceSet sourceURI(String sourceURI) {
        this.sourceURI = sourceURI;
        return this;
    }

    /**
     * Specifies a FASTA format file/string.
     *
     * @return sourceURI
     **/
    @Schema(example = "https://wiki.brapi.org/files/demo.fast", description = "Specifies a FASTA format file/string.")
    public String getSourceURI() {
        return sourceURI;
    }

    public void setSourceURI(String sourceURI) {
        this.sourceURI = sourceURI;
    }

    public ReferenceSet species(OntologyTerm species) {
        this.species = species;
        return this;
    }

    /**
     * Get species
     *
     * @return species
     **/
    @Schema(description = "")
    public OntologyTerm getSpecies() {
        return species;
    }

    public void setSpecies(OntologyTerm species) {
        this.species = species;
    }


    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        ReferenceSet referenceSet = (ReferenceSet) o;
        return Objects.equals(this.additionalInfo, referenceSet.additionalInfo) &&
                Objects.equals(this.assemblyPUI, referenceSet.assemblyPUI) &&
                Objects.equals(this.commonCropName, referenceSet.commonCropName) &&
                Objects.equals(this.description, referenceSet.description) &&
                Objects.equals(this.externalReferences, referenceSet.externalReferences) &&
                Objects.equals(this.isDerived, referenceSet.isDerived) &&
                Objects.equals(this.md5checksum, referenceSet.md5checksum) &&
                Objects.equals(this.referenceSetDbId, referenceSet.referenceSetDbId) &&
                Objects.equals(this.referenceSetName, referenceSet.referenceSetName) &&
                Objects.equals(this.sourceAccessions, referenceSet.sourceAccessions) &&
                Objects.equals(this.sourceGermplasm, referenceSet.sourceGermplasm) &&
                Objects.equals(this.sourceURI, referenceSet.sourceURI) &&
                Objects.equals(this.species, referenceSet.species);
    }

    @Override
    public int hashCode() {
        return Objects.hash(additionalInfo, assemblyPUI, commonCropName, description, externalReferences, isDerived, md5checksum, referenceSetDbId, referenceSetName, sourceAccessions, sourceGermplasm, sourceURI, species);
    }


    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("class ReferenceSet {\n");

        sb.append("    additionalInfo: ").append(toIndentedString(additionalInfo)).append("\n");
        sb.append("    assemblyPUI: ").append(toIndentedString(assemblyPUI)).append("\n");
        sb.append("    commonCropName: ").append(toIndentedString(commonCropName)).append("\n");
        sb.append("    description: ").append(toIndentedString(description)).append("\n");
        sb.append("    externalReferences: ").append(toIndentedString(externalReferences)).append("\n");
        sb.append("    isDerived: ").append(toIndentedString(isDerived)).append("\n");
        sb.append("    md5checksum: ").append(toIndentedString(md5checksum)).append("\n");
        sb.append("    referenceSetDbId: ").append(toIndentedString(referenceSetDbId)).append("\n");
        sb.append("    referenceSetName: ").append(toIndentedString(referenceSetName)).append("\n");
        sb.append("    sourceAccessions: ").append(toIndentedString(sourceAccessions)).append("\n");
        sb.append("    sourceGermplasm: ").append(toIndentedString(sourceGermplasm)).append("\n");
        sb.append("    sourceURI: ").append(toIndentedString(sourceURI)).append("\n");
        sb.append("    species: ").append(toIndentedString(species)).append("\n");
        sb.append("}");
        return sb.toString();
    }

    /**
     * Convert the given object to string with each line indented by 4 spaces
     * (except the first line).
     */
    private String toIndentedString(Object o) {
        if (o == null) {
            return "null";
        }
        return o.toString().replace("\n", "\n    ");
    }

}
